### ISSUE TITLE
Issue binding Sprint to the R key

### ISSUE THREAD

**Issue by Aya-Komichi** [minecraft-access/minecraft-access#310]
When running version 1.4.3 of this mod with minecraft 1.20.1 on fabric I find myself unable to bind the R key to the sprint action. I double checked this by disabling the mod and found out I could then use the custom keybind as expected. I am not playing the latest version since I am playing another mod which only supports minecraft versions as recent as 1.20.2 (a version not supported by any other version of this mod btw) and I needed the mouse simulation functionality.

I already looked at the logs of the minecraft instance I ran this on, but I could not find anything obvious that gives any context to this problem. I can still try to include a log if you so desire. I did look at the default keybinds of this mod and found that R is bound to the health and hunger speech functionality, so I am assuming for now that specific detail has something to do with this issue.

I understand if you decide not to fix this issue for this slightly outdated mod version or if you decide not to include support for earlier 1.20.x versions on a newer mod version, but I would at least like to hear your response in that case.

**Comment by boholder** [comment]
First, thanks for this high quality detailed issue.
What's confused me is the keystroke checking logic isn't preemptive (from what I know), single key should be able to be checked several times per tick, and all corresponding operations will be performed.
An alternative solution is to remove or change this mod's `Speak Player Health and Hunger Key`, I think it would work. If that works, I'd prefer not to fix it at 1.4.3... thanks for understanding that.
But this does is a problem, so I'll try to reproduce and fix it in future versions.
If anything else needed, please keep in touch.

**Comment by Aya-Komichi** [comment]
> An alternative solution is to remove or change this mod's Speak Player Health and Hunger Key, I think it would work.

Oh I wish it had been that easy, because I tried both things, it does not fix the issue. What is stranger is I can bind Sprint to many other keys and they will just work, however the other keys aren't as comfortable to use. And I did the configuration so nothing else is bound to R really.

My assumption on the default hunger keybind causing the problem could be wrong because of this detail, but I really do not know what else can be the cause since I do not know the code. ¯\\\_(ツ)_/¯

**Comment by Aya-Komichi** [comment]
Another update. I decided to try and bind other things besides Sprint to the R key to see what would happen, and I found some interesting results.

Binding Walk, Off-Hand Swap, Inventory or Open Chat to R makes them no longer work. This is not exhaustive obviously but it should give you an idea of what's going on. Funnily enough, when I tried to bind Advancements to R, it just worked anyway so I have no idea what's exactly going on with this mod.

**Comment by boholder** [comment]
Found the problem.
It turns out that keystroke **setting (not checking)** logic (in original code) is preemptive, located in `KeyBinding.onKeyPressed()`:

```java
public static void onKeyPressed(InputUtil.Key key) {
        KeyBinding keyBinding = (KeyBinding)KEY_TO_BINDINGS.get(key); <---- one key for one KeyBinding instance
...
}
```

And the `KEY_TO_BINDINGS` will be updated when **initializing (not registering)** KeyBinding instances.
And... there is a unfinished feature that has an keybinding which default value is, you get it, `R`!

https://github.com/khanshoaib3/minecraft-access/blob/2d957f12701779e2f370d94b4f680cc51229895c/common/src/main/java/com/github/khanshoaib3/minecraft_access/utils/KeyBindingsHandler.java#L387-L392

I kept these keybindings unregistered, not realized that it's not enough.
So the result is when you set original sprint key to `R`, and press `R`, the `KeyBinding.onKeyPressed()` will set the unused key instead of sprint key as pressed, thus the sprint checking condition (located in `ClientPlayerEntity.tickMovement()`) is always false:

```java
        if ((bl6 || this.isSubmergedInWater()) && bl7 && bl5) {
            if (this.ticksLeftToDoubleTapSprint <= 0 && !this.client.options.sprintKey.isPressed()) {
                this.ticksLeftToDoubleTapSprint = 7;
            } else {
                this.setSprinting(true);
            }
        }
```

Why this mod's custom keybindings won't collision because we're using custom key-pressed checking logic instead of `KeyBinding` thing.

**Comment by boholder** [comment]
@Aya-Komichi Nope, yesterday fixing isn't solve the problem, wait a minute ~~Here, take this newly built file, GitHub doesn't allow uploading jar file so you need to remove the `.txt` suffix.
[minecraft-access-1.4.3+1.20.1-fabric.jar.txt](https://github.com/user-attachments/files/16480725/minecraft-access-1.4.3%2B1.20.1-fabric.jar.txt)~~

You can fork this branch and run the build action on github, in case you want to make some modifications: https://github.com/boholder/minecraft-access/tree/v1.4.3-1.20.1

**Comment by Aya-Komichi** [comment]
> @Aya-Komichi Nope, yesterday fixing isn't solve the problem, wait a minute ~Here, take this newly built file, GitHub doesn't allow uploading jar file so you need to remove the `.txt` suffix. [minecraft-access-1.4.3+1.20.1-fabric.jar.txt](https://github.com/user-attachments/files/16480725/minecraft-access-1.4.3%2B1.20.1-fabric.jar.txt)~
> 
> You can fork this branch and run the build action on github, in case you want to make some modifications: https://github.com/boholder/minecraft-access/tree/v1.4.3-1.20.1

Well, I tested the jar you sent me and I am happy to report it works as it's supposed to for my use case. Do keep in mind that I only used the mouse simulation, so I did not test all the other features and honestly, I don't have time and will to do so, so if you have to fix these bugs in newer versions, you should at least test if everything is working.

As for the branch, I guess I could some day try to make changes to it to make it build for 1.20.2 if possible at some point. I am not sure if I should close this issue yet since you may or may not have to patch the issues in newer versions which may or may not take some more time. However, you may close this now if you wish.

**Comment by boholder** [comment]
> I tested the jar you sent me and I am happy to report it works as it's supposed to for my use case.

@Aya-Komichi I forgot to send the new jar file today, so I'm surprised that the jar yesterday does work, here:
[minecraft-access-1.4.3+1.20.1-fabric.jar.txt](https://github.com/user-attachments/files/16487347/minecraft-access-1.4.3%2B1.20.1-fabric.jar.txt)
I also changed the jar file in GitHub release.

> since you may or may not have to patch the issues in newer versions

Bug fixed today with this [commit](https://github.com/boholder/minecraft-access/commit/e1cd90fe6c51893adc0a3fe9f144af33dc4950a0), test case is sprint bind to R key.

**Comment by boholder** [comment]
Also updated jar files on Modrinth and CurseForge, I think it's over?

**Comment by Aya-Komichi** [comment]
> Also updated jar files on Modrinth and CurseForge, I think it's over?

Yeah, I believe it is over, I played last night with the newer jar you sent, it worked well and I had fun.

One curiosity, if I decided to build this for version 1.20.2 of minecraft would it just be a question of updating the dependencies or is there any area of the code that needs to be changed? If you aren't sure, it's fine.

**Comment by boholder** [comment]
> would it just be a question of updating the dependencies or is there any area of the code that needs to be changed? 

Both situations would be possible, depending on whether mojang changed code in 1.20.2. You can check this mod's upgrade commits to know what needs to be changed.

**Comment by Aya-Komichi** [comment]
> Both situations would be possible, depending on whether mojang changed code in 1.20.2. You can check this mod's upgrade commits to know what needs to be changed.

Well, I gave it a try. I picked dependency versions that were compatible, but tried not to pick the most up to date ones and I fixed the errors that appeared by copying the code changes on the git diff shown by github in comparison with version 1.5.0. In the end the mod jar builds just fine but I get a lot of exceptions when trying to launch a minecraft instance with it and other mods.

I'm not a minecraft mod developer and honestly the way Microsoft is slowly ruining minecraft these days doesn't encourage me to make more effort than this. I will just be content with what I already got. Thank you for your help.

**Comment by boholder** [comment]
I don't think 1.20.2 has new content comparing to 1.20.1, if this can comfort you.
https://minecraft.wiki/w/Java_Edition_1.20.2

**Comment by Aya-Komichi** [comment]
@boholder I start to wonder if the problem I was having was building the jar itself, today I tried to make a different and simpler change and no matter what the jar fails with a similar output. Taking a look at the jar I finally check that the manifest is junk.

I have been trying to build the jar in IntelliJ using the shadowjar task for fabric, but that somehow doesn't seem to be the correct way or something else is missing? What is the recommended way to build the jars of this mod?

**Comment by boholder** [comment]
`gradlew fabric:build`
See https://github.com/khanshoaib3/minecraft-access/blob/1.21/CONTRIBUTING.md#setting-up-local-environment

**Comment by Aya-Komichi** [comment]
> `gradlew fabric:build` See https://github.com/khanshoaib3/minecraft-access/blob/1.21/CONTRIBUTING.md#setting-up-local-environment

Thanks a lot, I guess that's what I get for not looking enough in the repo for more md files. Now I actually can build the jars properly. I probably could build 1.4.3 for 1.20.2 now, but I'm actually not that interested on that anymore, even thou I have those changes saved.

This time since I have been looking to make a working wayland setup for myself and there doesn't seem to be any progress on the support for ydotool yet, I decided to look more less how the mod was using the xdotool and when I found the command strings I decided to just edit them. I did some testing with ydotool to try and get a behaviour that resembles what the mod should offer and I got mixed results:
- Mouse clicking actually seems to just work with this, would be a shame it didn't after all the testing I had done.
- I tried testing mouse movement with the inventory controls even thou I don't need them, I don't know exactly how their beahviour looks like with the xdotool tbh, but it didn't seem to do anything with the ydotool commands I configured.

I actually only tested the mouse movement ydotool commands on a terminal after building the jar and they should definitely do something, but I found out they behave weirdly. My monitor has a resolution of 1920x1080, yet this seems to send the cursor to the bottom right corner with half of that resolution passed into the arguments. I have to review what xdotool does, but to me this doesn't make a lot of sense yet.

This should be my final reply here since further talking on this ydotool topic should be done in the open ydotool support issue instead of this already closed issue. Not really sure if I will pursue that thou, just because the ydotool behaves like this on my system, doesn't mean it will do it in other systems and I can't test that myself. Also the setup of the ydotool is tricky compared with the xdotool, so it could be too early to get ydotool support officially.
